File: java/spatiad-sdk/src/main/java/com/spatiad/sdk/SpatiadClient.java
package com.spatiad.sdk;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;

/**
 * Lightweight HTTP client for the Spatiad spatial dispatch engine REST API.
 *
 * <p>Uses Java 11+ {@link HttpClient} with Jackson for JSON serialization.
 * Retries failed requests with exponential backoff (3 attempts, 150ms base).
 */
public class SpatiadClient {

    private static final int MAX_ATTEMPTS = 3;
    private static final long BASE_BACKOFF_MS = 150;

    private final String baseUrl;
    private final String dispatcherToken;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    /**
     * Create a client without authentication.
     *
     * @param baseUrl the Spatiad server base URL (e.g. "http://localhost:8080")
     */
    public SpatiadClient(String baseUrl) {
        this(baseUrl, null);
    }

    /**
     * Create a client with a dispatcher bearer token.
     *
     * @param baseUrl          the Spatiad server base URL
     * @param dispatcherToken  optional bearer token (may be null)
     */
    public SpatiadClient(String baseUrl, String dispatcherToken) {
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.dispatcherToken = dispatcherToken;
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();
        this.objectMapper = new ObjectMapper()
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                .setSerializationInclusion(JsonInclude.Include.NON_NULL);
    }

    /**
     * Create a dispatch offer.
     */
    public CreateOfferResponse createOffer(CreateOfferRequest request) {
        return post("/api/v1/dispatch/offer", request, CreateOfferResponse.class);
    }

    /**
     * Insert or update a driver's location and status.
     */
    public void upsertDriver(UpsertDriverRequest request) {
        post("/api/v1/driver/upsert", request, Void.class);
    }

    /**
     * Cancel an active offer.
     */
    public void cancelOffer(String offerId) {
        post("/api/v1/dispatch/cancel", Map.of("offer_id", offerId), Void.class);
    }

    /**
     * Cancel a job.
     */
    public void cancelJob(String jobId) {
        post("/api/v1/dispatch/job/cancel", Map.of("job_id", jobId), Void.class);
    }

    /**
     * Get the current status of a job.
     */
    public JobStatusResponse getJobStatus(String jobId) {
        return get("/api/v1/dispatch/job/" + encode(jobId), JobStatusResponse.class);
    }

    /**
     * Get events for a job with optional pagination and filtering.
     */
    public JobEventsResponse getJobEvents(GetJobEventsRequest request) {
        StringJoiner query = new StringJoiner("&");
        if (request.getLimit() != null) {
            query.add("limit=" + request.getLimit());
        }
        if (request.getCursor() != null) {
            query.add("cursor=" + encode(request.getCursor()));
        }
        List<String> kinds = request.getKinds();
        if (kinds != null && !kinds.isEmpty()) {
            for (String kind : kinds) {
                query.add("kinds=" + encode(kind));
            }
        }

        String path = "/api/v1/dispatch/job/" + encode(request.getJobId()) + "/events";
        if (query.length() > 0) {
            path += "?" + query;
        }
        return get(path, JobEventsResponse.class);
    }

    // ---- internal helpers ----

    private <T> T post(String path, Object body, Class<T> responseType) {
        try {
            byte[] jsonBody = objectMapper.writeValueAsBytes(body);
            HttpRequest.Builder builder = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofByteArray(jsonBody));
            applyAuth(builder);
            return executeWithRetry(builder.build(), responseType);
        } catch (SpatiadApiError e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Failed to execute POST " + path, e);
        }
    }

    private <T> T get(String path, Class<T> responseType) {
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .GET();
            applyAuth(builder);
            return executeWithRetry(builder.build(), responseType);
        } catch (SpatiadApiError e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Failed to execute GET " + path, e);
        }
    }

    private void applyAuth(HttpRequest.Builder builder) {
        if (dispatcherToken != null && !dispatcherToken.isEmpty()) {
            builder.header("Authorization", "Bearer " + dispatcherToken);
        }
    }

    private <T> T executeWithRetry(HttpRequest request, Class<T> responseType)
            throws IOException, InterruptedException {
        IOException lastException = null;

        for (int attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
            if (attempt > 0) {
                long backoff = BASE_BACKOFF_MS * (1L << (attempt - 1));
                Thread.sleep(backoff);
            }
            try {
                HttpResponse<byte[]> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofByteArray());

                int status = response.statusCode();
                if (status >= 200 && status < 300) {
                    if (responseType == Void.class || response.body().length == 0) {
                        return null;
                    }
                    return objectMapper.readValue(response.body(), responseType);
                }

                if (status >= 500 || status == 429) {
                    lastException = new IOException("HTTP " + status);
                    continue;
                }

                // Client error -- do not retry
                String code = null;
                String message = null;
                try {
                    ErrorBody err = objectMapper.readValue(response.body(), ErrorBody.class);
                    code = err.code;
                    message = err.message;
                } catch (Exception ignored) {
                    message = response.body().length > 0
                            ? new String(response.body(), StandardCharsets.UTF_8)
                            : null;
                }
                throw new SpatiadApiError(status, code, message);

            } catch (IOException e) {
                lastException = e;
            }
        }

        throw lastException != null ? lastException : new IOException("Request failed after retries");
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    /** Internal DTO for parsing error response bodies. */
    private static class ErrorBody {
        @JsonProperty("code")
        public String code;
        @JsonProperty("message")
        public String message;
    }
}

File: kotlin/spatiad-sdk/src/main/kotlin/com/spatiad/sdk/SpatiadClient.kt
package com.spatiad.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Lightweight HTTP client for the Spatiad spatial dispatch REST API.
 *
 * All public methods are `suspend` functions safe to call from any coroutine
 * context; network I/O is dispatched on [Dispatchers.IO].
 *
 * ```kotlin
 * val client = SpatiadClient(baseUrl = "https://api.spatiad.example.com", token = "sk-...")
 * val response = client.createOffer(CreateOfferRequest(...))
 * client.close()
 * ```
 */
class SpatiadClient(
    private val baseUrl: String,
    private val token: String? = null,
    private val maxRetries: Int = 3,
    private val retryBaseDelayMs: Long = 150L,
    private val retryableStatuses: Set<Int> = setOf(408, 429, 500, 502, 503, 504),
    httpClient: OkHttpClient? = null,
) {
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
    }

    private val client: OkHttpClient = httpClient ?: OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val normalizedBaseUrl: String = baseUrl.trimEnd('/')

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    // -- public API ----------------------------------------------------------

    /**
     * Creates a dispatch offer.
     *
     * POST /api/v1/dispatch/offer
     */
    suspend fun createOffer(request: CreateOfferRequest): CreateOfferResponse {
        val body = json.encodeToString(request)
        val response = executeWithRetry(
            buildPost("/api/v1/dispatch/offer", body),
        )
        return json.decodeFromString(response)
    }

    /**
     * Creates or updates a driver's status and position.
     *
     * POST /api/v1/driver/upsert
     */
    suspend fun upsertDriver(request: UpsertDriverRequest) {
        val body = json.encodeToString(request)
        executeWithRetry(buildPost("/api/v1/driver/upsert", body))
    }

    /**
     * Cancels an active offer.
     *
     * POST /api/v1/dispatch/cancel
     */
    suspend fun cancelOffer(offerId: String) {
        val body = json.encodeToString(CancelOfferBody(offerId))
        executeWithRetry(buildPost("/api/v1/dispatch/cancel", body))
    }

    /**
     * Cancels an active job.
     *
     * POST /api/v1/dispatch/job/cancel
     */
    suspend fun cancelJob(jobId: String) {
        val body = json.encodeToString(CancelJobBody(jobId))
        executeWithRetry(buildPost("/api/v1/dispatch/job/cancel", body))
    }

    /**
     * Returns the current status of a job.
     *
     * GET /api/v1/dispatch/job/{job_id}
     */
    suspend fun getJobStatus(jobId: String): JobStatusResponse {
        val response = executeWithRetry(buildGet("/api/v1/dispatch/job/$jobId"))
        return json.decodeFromString(response)
    }

    /**
     * Returns events for a job with optional pagination and filtering.
     *
     * GET /api/v1/dispatch/job/{job_id}/events
     */
    suspend fun getJobEvents(request: GetJobEventsRequest): JobEventsResponse {
        val urlBuilder = "$normalizedBaseUrl/api/v1/dispatch/job/${request.jobId}/events"
            .toHttpUrl()
            .newBuilder()

        request.limit?.let { urlBuilder.addQueryParameter("limit", it.toString()) }
        request.cursor?.let { urlBuilder.addQueryParameter("cursor", it) }
        request.kinds?.takeIf { it.isNotEmpty() }?.let {
            urlBuilder.addQueryParameter("kinds", it.joinToString(","))
        }

        val httpRequest = newRequestBuilder()
            .url(urlBuilder.build())
            .get()
            .build()

        val response = executeWithRetry(httpRequest)
        return json.decodeFromString(response)
    }

    /**
     * Shuts down the underlying OkHttp connection pool and dispatcher.
     */
    fun close() {
        client.dispatcher.executorService.shutdown()
        client.connectionPool.evictAll()
    }

    // -- internals -----------------------------------------------------------

    private fun newRequestBuilder(): Request.Builder {
        val builder = Request.Builder()
        token?.let { builder.addHeader("Authorization", "Bearer $it") }
        return builder
    }

    private fun buildPost(path: String, jsonBody: String): Request =
        newRequestBuilder()
            .url("$normalizedBaseUrl$path")
            .post(jsonBody.toRequestBody(jsonMediaType))
            .build()

    private fun buildGet(path: String): Request =
        newRequestBuilder()
            .url("$normalizedBaseUrl$path")
            .get()
            .build()

    /**
     * Executes [request] with exponential-backoff retries on transient errors.
     *
     * Returns the response body as a [String]. Throws [SpatiadApiError] on
     * non-success responses that are not retryable (or after all retries are
     * exhausted), and re-throws [IOException] for transport-level failures.
     */
    private suspend fun executeWithRetry(request: Request): String {
        var lastException: Exception? = null

        for (attempt in 0 until maxRetries) {
            try {
                val response = execute(request)
                val responseBody = response.body?.string().orEmpty()

                if (response.isSuccessful) {
                    return responseBody
                }

                val apiError = parseError(response.code, responseBody, request)

                if (response.code !in retryableStatuses || attempt == maxRetries - 1) {
                    throw apiError
                }

                lastException = apiError
            } catch (e: IOException) {
                lastException = e
                if (attempt == maxRetries - 1) throw e
            }

            delay(retryBaseDelayMs * (1L shl attempt))
        }

        throw lastException ?: SpatiadApiError(
            statusCode = 0,
            message = "Request failed after $maxRetries retries",
        )
    }

    private suspend fun execute(request: Request): Response =
        withContext(Dispatchers.IO) {
            client.newCall(request).execute()
        }

    private fun parseError(
        statusCode: Int,
        body: String,
        request: Request,
    ): SpatiadApiError {
        val parsed = try {
            json.decodeFromString<ApiErrorBody>(body)
        } catch (_: Exception) {
            null
        }

        return SpatiadApiError(
            statusCode = statusCode,
            code = parsed?.error,
            message = parsed?.message
                ?: "${request.method} ${request.url.encodedPath} failed with status $statusCode",
        )
    }
}
